src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -266,5 +273,33 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
                 builder.Clear();
             }
         }
+
+        private string FastGetUrl(
+            string protocol, 
+            string host, 
+            VirtualPathData pathData, 
+            string fragment)
+        {
+            if (string.IsNullOrEmpty(protocol) 
+                && string.IsNullOrEmpty(host) 
+                && string.IsNullOrEmpty(fragment))
```

Comment:
what happens today when these are empty? Generally we don't treat empty and null as equivalent unless they really are - or we have a legacy/compatibility requirement.